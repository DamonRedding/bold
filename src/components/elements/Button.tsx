import * as React from 'react'
import * as classnames from 'classnames'
import { helpersClassnames, filterProps, UtilProps } from '../../util/Util'
import { Icon } from './Icon'
import withHint, { WithHintProps } from '../decorators/withHint'

export interface ButtonProps extends UtilProps {
    className?: string
    dashed?: boolean
    disabled?: boolean
    icon?: string
    loading?: boolean
    name?: string,
    onClick?: Function
    outlined?: boolean
    shadow?: boolean
    size?: string
    square?: boolean
    tabIndex?: number
    title?: string
    type?: string
}

export interface ButtonState {
    loading: boolean
}

export class ButtonComponent extends React.Component<ButtonProps, ButtonState> {

    private timeout: number

    constructor() {
        super()
        this.handleOnKeyPress = this.handleOnKeyPress.bind(this)
        this.onClick = this.onClick.bind(this)

        this.state = {
            loading: false
        }
    }

    render() {
        const classes: string = classnames('button', this.props.className, helpersClassnames(this.props), {
            'is-grey': this.props.type && this.props.type === 'grey',
            'is-primary': this.props.type && this.props.type === 'primary',
            'is-transparent': this.props.type && this.props.type === 'transparent',
            'is-neon': this.props.type && this.props.type === 'neon',
            'is-success': this.props.type && this.props.type === 'success',
            'is-danger': this.props.type && this.props.type === 'danger',
            'is-warning': this.props.type && this.props.type === 'warning',
            'is-info': this.props.type && this.props.type === 'info',
            'is-link': this.props.type && this.props.type === 'link',
            'is-small': this.props.size && this.props.size === 'small',
            'is-normal': this.props.size && this.props.size === 'normal',
            'is-medium': this.props.size && this.props.size === 'medium',
            'is-large': this.props.size && this.props.size === 'large',
            'is-dashed': this.props.dashed,
            'is-loading': this.props.loading || this.state.loading,
            'is-outlined': this.props.outlined,
            'is-square': this.props.square,
            'has-shadow': this.props.shadow,
        })

        return (
            <button
                {...filterProps(this.props, ...excludedProps) }
                className={classes}
                name={this.props.name}
                onClick={this.onClick}
                onKeyPress={this.handleOnKeyPress}
                tabIndex={this.props.tabIndex}
                title={this.props.title}
                type='button'
            >
                {this.props.icon && <span className='icon'><Icon icon={this.props.icon} /></span>}
                {this.props.children && <span>{this.props.children}</span>}
            </button>
        )
    }

    private handleOnKeyPress(event) {
        event.preventDefault()
        if (!event) { event = window.event } // cross-browser shenanigans
        if (event.charCode === 32 || event.charCode === 13 && this.props.onClick) { // this is the spacebar
            this.onClick(event)
        }
        return true // treat all other keys normally;
    }

    componentWillUnmount() {
        clearTimeout(this.timeout)
        this.timeout = -1
    }

    private onClick(event) {
        if (this.props.onClick) {

            const promise = this.props.onClick(event)
            if (promise && promise.then) {
                this.setState({ loading: true })
                promise
                    .then(() => this.stopLoading())
                    .catch((error) => {
                        this.stopLoading()
                        throw new Error(error)
                    })
            }
        }
    }

    private stopLoading() {
        if (!this.timeout) {
            this.timeout = window.setTimeout(() => {
                this.setState({ loading: false })
                this.timeout = null
            }, 10)
        }
    }

}

export const Button: React.ComponentClass<WithHintProps & ButtonProps> = withHint(ButtonComponent)


const excludedProps = [
    'dashed',
    'icon',
    'loading',
    'outlined',
    'shadow',
    'square',
    'size',
    'type',
]
