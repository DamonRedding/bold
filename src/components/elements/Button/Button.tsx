import * as React from 'react'
import * as classnames from 'classnames'
import { helpersClassnames, UtilProps } from '../../../util/Util'
import { Icon } from '../Icon'
import withHint, { WithHintProps } from '../../decorators/withHint'
import { withStyles, WithStylesProps, css } from '../../decorators/withStyles'
import { MouseEventHandler } from 'react'

export type Type = 'success' | 'grey' | 'primary' | 'transparent' | 'neon' | 'danger' | 'warning' | 'info' | 'link'

export interface ButtonProps extends UtilProps, WithHintProps, WithStylesProps {
    className?: string
    dashed?: boolean
    disabled?: boolean
    icon?: string
    loading?: boolean
    name?: string,
    onClick?: Function
    onMouseEnter?: MouseEventHandler<any>
    onMouseLeave?: MouseEventHandler<any>
    outlined?: boolean
    shadow?: boolean
    size?: string
    square?: boolean
    tabIndex?: number
    title?: string
    type?: Type
}

export interface ButtonState {
    loading: boolean
}

@withStyles
@withHint
export class Button extends React.Component<ButtonProps, ButtonState> {

    private timeout: number

    constructor(props, context?) {
        super(props, context)

        this.state = {
            loading: false
        }
    }

    render() {
        const styles = this.props.createStyles(theme => ({
            button: {
                borderRadius: 100,
                cursor: 'pointer',
                fontSize: 13,
                fontWeight: 600,
                height: 34,
                paddingLeft: 20,
                paddingRight: 20
            },
            primary: {
                backgroundColor: theme.primary,
                border: '1px solid' + theme.primary,
                color: 'white'
            }
        }))

        const classes = css(
            styles.button,
            this.props.type === 'primary' && styles.primary
        )
        classnames('button', this.props.className, helpersClassnames(this.props), {
            'is-grey': this.props.type && this.props.type === 'grey',
            'is-primary': this.props.type && this.props.type === 'primary',
            'is-transparent': this.props.type && this.props.type === 'transparent',
            'is-neon': this.props.type && this.props.type === 'neon',
            'is-success': this.props.type && this.props.type === 'success',
            'is-danger': this.props.type && this.props.type === 'danger',
            'is-warning': this.props.type && this.props.type === 'warning',
            'is-info': this.props.type && this.props.type === 'info',
            'is-link': this.props.type && this.props.type === 'link',
            'is-small': this.props.size && this.props.size === 'small',
            'is-normal': this.props.size && this.props.size === 'normal',
            'is-medium': this.props.size && this.props.size === 'medium',
            'is-large': this.props.size && this.props.size === 'large',
            'is-dashed': this.props.dashed,
            'is-loading': this.props.loading || this.state.loading,
            'is-outlined': this.props.outlined,
            'is-square': this.props.square,
            'has-shadow': this.props.shadow,
        })

        return (
            <button
                className={classes}
                name={this.props.name}
                onClick={this.onClick}
                onKeyPress={this.handleOnKeyPress}
                onMouseEnter={this.props.onMouseEnter}
                onMouseLeave={this.props.onMouseLeave}
                tabIndex={this.props.tabIndex}
                title={this.props.title}
                type='button'
            >
                {this.props.icon && <span className='icon'><Icon icon={this.props.icon} /></span>}
                {this.props.children && <span>{this.props.children}</span>}
            </button>
        )
    }

    private handleOnKeyPress = (event) => {
        event.preventDefault()
        if (!event) { event = window.event } // cross-browser shenanigans
        if (event.charCode === 32 || event.charCode === 13 && this.props.onClick) { // this is the spacebar
            this.onClick(event)
        }
        return true // treat all other keys normally;
    }

    componentWillUnmount() {
        clearTimeout(this.timeout)
        this.timeout = -1
    }

    private onClick = (event) => {
        if (this.props.onClick) {

            const promise = this.props.onClick(event)
            if (promise && promise.then) {
                this.setState({ loading: true })
                promise
                    .then(() => this.stopLoading())
                    .catch((error) => {
                        this.stopLoading()
                        throw new Error(error)
                    })
            }
        }
    }

    private stopLoading() {
        if (!this.timeout) {
            this.timeout = window.setTimeout(() => {
                this.setState({ loading: false })
                this.timeout = null
            }, 10)
        }
    }

}
