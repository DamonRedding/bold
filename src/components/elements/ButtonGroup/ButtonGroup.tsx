import { Interpolation } from 'emotion'
import * as React from 'react'

import { Styles, withStyles, WithStylesProps } from '../../../styles/withStyles'

export interface ButtonGroupProps extends WithStylesProps {
    style?: Interpolation
}

@withStyles
export class ButtonGroup extends React.PureComponent<ButtonGroupProps> {
    render() {
        const { children, css, style } = this.props
        const styles: Styles = {
            root: {
                '& > button:not(:first-child)': {
                    borderTopLeftRadius: '0',
                    borderBottomLeftRadius: '0',
                },
                '& > button:not(:last-child)': {
                    borderTopRightRadius: '0',
                    borderBottomRightRadius: '0',
                },
                '& > button:focus': {
                    zIndex: 1, // prevent box-shadow overlapping
                },
            },
        }

        return (
            <div className={css(styles.root, style)} role='group'>
                {children}
            </div>
        )
    }
}
