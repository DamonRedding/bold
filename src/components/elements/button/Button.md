# Button

Botões expressam que alguma atividade vai acontecer quando o usuário clicar nele.
São usados para iniciar uma ação, alguns exemplos incluem a ação de Adicionar, Salvar, Deletar.

## Boas práticas

- Não use botões como elementos de navegação. Use Links ao invés,  porque levam o usuário a uma
nova página que não está associada com a ação.
- A label do botão indica ao usuário o que vai acontecer quando ele clicar no botão.
Use somente verbos que descrevem a ação, como Adicionar ou Deletar. As labels devem seguir o
padrão Capitalized (apenas a primeira letra na palavra em maiúsculo).
