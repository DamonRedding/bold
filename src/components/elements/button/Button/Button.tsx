import * as React from 'react'

import { focusBoxShadow, shade, Theme, withStyles, WithStylesProps } from '../../../../styles'
import { Icons } from '../../Icon/generated/Icons'
import { Icon } from '../../Icon/Icon'
import { BaseButton, BaseButtonProps } from '../BaseButton'

export type Type = 'normal' | 'primary'

export interface ButtonProps extends BaseButtonProps, WithStylesProps {
    icon?: Icons
    label: string
    loading?: boolean
    type?: Type
}

export interface ButtonState {
    loading: boolean
}

export const createStyles = (theme: Theme) => ({
    button: {
        backgroundColor: theme.color.white,
        border: '1px solid ' + theme.color.gray70,
        borderRadius: theme.baseRadius + 2,
        color: theme.color.gray40,
        cursor: 'pointer',
        display: 'inline-block',
        fontFamily: theme.font.textFamily,
        position: 'relative',
        userSelect: 'none',
        fontSize: '0.875rem',
        fontWeight: 'bold',
        letterSpacing: 1,
        padding: '1rem 2.5rem',
        transition: 'all .2s',
        transform: 'translate3d(0,0,0)',
        'span': {
            transition: 'color .2s',
        },
        ':not(:disabled):active': {
            boxShadow: 'inset 0 2px 8px 0 rgba(0, 0, 0, 0.1)',
        },
        ':disabled': {
            opacity: 0.5,
        },
        ':focus': {
            outline: 'none',
            boxShadow: focusBoxShadow(theme),
        },
        ':hover': {
            backgroundColor: shade(-0.08, theme.color.white),
        },
    },
    primary: {
        backgroundColor: theme.color.primary,
        border: '1px solid ' + theme.color.primary,
        color: theme.color.white,
        ':hover': {
            backgroundColor: shade(-0.08, theme.color.primary),
        },
    },
    loading: {
        pointerEvents: 'none',
        'span': {
            color: 'transparent',
        },
        ':after': {
            animation: `${theme.animation.spinAround} 500ms infinite linear`,
            border: '3px solid currentColor',
            borderRadius: '50%',
            borderRightColor: 'transparent',
            borderTopColor: 'transparent',
            content: '""',
            display: 'block',
            height: '1.5em',
            width: '1.5em',
            position: 'absolute',
            left: 'calc(50% - (1.5em / 2))',
            top: 'calc(50% - (1.5em / 2))',
        },
    },
})

@withStyles
export class Button extends React.Component<ButtonProps, ButtonState> {

    static defaultProps: Partial<ButtonProps> = {
        type: 'normal',
    }

    constructor(props, context?) {
        super(props, context)

        this.state = {
            loading: false,
        }
    }

    render() {
        const {
            label,
            css,
            icon,
            loading,
            theme,
            type,
            ...rest,
        } = this.props

        const styles = createStyles(theme)

        const classes = css(
            styles.button,
            type === 'primary' && styles.primary,
            (this.state.loading || loading) && styles.loading
        )

        return (
            <BaseButton
                {...rest}
                className={classes}
                onLoadingChange={this.onLoadingChange}
            >
                <span>
                    {icon && <Icon icon={icon} />}
                    {label}
                </span>
            </BaseButton>
        )
    }

    private onLoadingChange = (loading: boolean) => {
        this.setState({ loading })
    }

}
