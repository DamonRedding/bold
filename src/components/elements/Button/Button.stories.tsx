import { action } from '@storybook/addon-actions'
import { boolean, select, text } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/react'
import * as React from 'react'
import { MemoryRouter } from 'react-router'

import * as Doc from './Button.md'
import { Button } from './Button/Button'
import { Size, Skins, Type } from './Button/ButtonSkins'
import { ButtonLink } from './ButtonLink/ButtonLink'

const typeOptions: { [key in Type]: string } = {
    'normal': 'normal',
    'primary': 'primary',
    'danger': 'danger',
}

const skinOptions: { [key in Skins]: string } = {
    'default': 'default',
    'ghost': 'ghost',
    'outline': 'outline',
}

const sizeOptions: { [key in Size]: string } = {
    'small': 'small',
    'medium': 'medium',
    'large': 'large',
}

storiesOf('Components/Button', module)
    // @ts-ignore
    .addParameters({
        info: { text: Doc },
    })
    .add('default', () => (
        <Button
            label={text('label', 'Button')}
            skin={select('skin', skinOptions, 'default')}
            size={select('size', sizeOptions, 'medium')}
            type={select('type', typeOptions, 'normal')}
            loading={boolean('loading', false)}
            disabled={boolean('disabled', false)}
            block={boolean('block', false)}
            onClick={action('button-clicked')}
        />
    ))
    .add('icon', () => (
        <Button
            icon='adjust'
            skin={select('skin', skinOptions, 'ghost')}
            size={select('size', sizeOptions, 'small')}
            type={select('type', typeOptions, 'normal')}
            loading={boolean('loading', false)}
            disabled={boolean('disabled', false)}
            block={boolean('block', false)}
            onClick={action('button-clicked')}
        />
    ))
    .add('link', () => (
        <MemoryRouter initialEntries={['/']} initialIndex={0}>
            <ButtonLink
                to='/links'
                icon='arrowLeft'
                label={text('label', 'Link')}
                skin={select('skin', skinOptions, 'default')}
                size={select('size', sizeOptions, 'medium')}
                type={select('type', typeOptions, 'normal')}
                loading={boolean('loading', false)}
                disabled={boolean('disabled', false)}
                block={boolean('block', false)}
                onClick={action('link-clicked')}
            />
        </MemoryRouter>
    ))
