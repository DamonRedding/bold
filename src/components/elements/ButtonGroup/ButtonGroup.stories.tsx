import { storiesOf } from '@storybook/react'
import * as React from 'react'

import { HFlow } from '../../layout/Flow/HFlow'
import { Button } from '../Button'
import { Icon } from '../Icon'

import { ButtonGroup } from './ButtonGroup'

storiesOf('Components/ButtonGroup', module)
    .add('default', () => (
        <HFlow>
            <ButtonGroup>
                <Button size='small' kind='primary'>First</Button>
                <Button size='small'>Second</Button>
                <Button size='small' disabled>Third</Button>
            </ButtonGroup>
            <ButtonGroup>
                <Button size='small'><Icon icon='trashOutline' /></Button>
                <Button size='small'><Icon icon='zoomOutline' /></Button>
            </ButtonGroup>
        </HFlow>
    ))
