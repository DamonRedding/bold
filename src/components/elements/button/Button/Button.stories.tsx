import { boolean, select, text, withKnobs } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/react'
import * as React from 'react'

import { withPropTypes } from '../../../../stories-addons/withPropTypes'
import { withTheme } from '../../../../stories-addons/withTheme'

import { Button, Type } from './Button'

const typeOptions: {[key in Type]: string} = {
    'normal': 'normal',
    'primary': 'primary',
}

storiesOf('Components', module)
    .addDecorator(withPropTypes(`
    Botões expressam que alguma atividade vai acontecer quando o usuário clicar nele.
    São usados para iniciar uma ação, alguns exemplos incluem a ação de Adicionar, Salvar, Deletar.

    ### BOAS PRÁTICAS:
    - Não use botões como elementos de navegação. Use Links ao invés,  porque levam o usuário a uma
    nova página que não está associada com a ação.
    - A label do botão indica ao usuário o que vai acontecer quando ele clicar no botão.
    Use somente verbos que descrevem a ação, como Adicionar ou Deletar. As labels devem seguir o
    padrão Capitalized (apenas a primeira letra na palavra em maiúsculo).
    `))
    .addDecorator(withKnobs)
    .addDecorator(withTheme())
    .add('Button', () => (
        <Button
            disabled={boolean('disabled', false)}
            type={select('type', typeOptions, 'normal')}
            label={text('label', 'Button')}
            loading={boolean('loading', false)}
        />
    ))
