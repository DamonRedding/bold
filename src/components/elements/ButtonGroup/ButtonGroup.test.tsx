import { mount, render } from 'enzyme'
import React from 'react'

import { withTheme } from '../../../test'
import { Button } from '../Button'

import { ButtonGroup } from './ButtonGroup'

it('should render correctly', () => {
    expect(render(withTheme(
        <ButtonGroup />
    ))).toMatchSnapshot()
})

it('should should accept stlye prop', () => {
    expect(render(withTheme(
        <ButtonGroup style={{ color: 'red' }} />
    ))).toMatchSnapshot()
})

it('should mount children', () => {
    const wrapper = mount(withTheme(
        <ButtonGroup>
            <Button>First</Button>
        </ButtonGroup>
    ))
    expect(wrapper.find(Button).length).toEqual(1)
})
