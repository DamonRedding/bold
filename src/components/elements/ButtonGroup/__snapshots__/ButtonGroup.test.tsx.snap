// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
.emotion-0 > button:not(:first-child),
.emotion-0 > a:not(:first-child) {
  border-top-left-radius: 0px;
  border-bottom-left-radius: 0px;
}

.emotion-0 > button:not(:last-child),
.emotion-0 > a:not(:last-child) {
  border-top-right-radius: 0px;
  border-bottom-right-radius: 0px;
}

.emotion-0 > button:focus,
.emotion-0 > a:focus {
  z-index: 1;
}

<div
  class="emotion-0"
  role="group"
/>
`;

exports[`should should accept stlye prop 1`] = `
.emotion-0 {
  color: red;
}

.emotion-0 > button:not(:first-child),
.emotion-0 > a:not(:first-child) {
  border-top-left-radius: 0px;
  border-bottom-left-radius: 0px;
}

.emotion-0 > button:not(:last-child),
.emotion-0 > a:not(:last-child) {
  border-top-right-radius: 0px;
  border-bottom-right-radius: 0px;
}

.emotion-0 > button:focus,
.emotion-0 > a:focus {
  z-index: 1;
}

<div
  class="emotion-0"
  role="group"
/>
`;
